File: _wpf/wpf-business-programmer/2-wpf-business-programmer-m2-exercise-files/Samples/CSharp/WPFListBox/WPFComponents/Converters/IndexToMarginConverter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using System.Globalization;
using System.ComponentModel;
using System.Windows.Data;
using System.Windows;
using System.Windows.Controls;

namespace WPFComponents
{
  [ValueConversion(typeof(ListBoxItem), typeof(Thickness))]
  public class IndexToMarginConverter : IValueConverter
  {
    public object Convert(object value, Type targetType, object parameter, CultureInfo culture)
    {
      Thickness ret;
      int itemIndex;
      ListBoxItem lstItem;
      ListBox lstBox;

      lstItem = (value as ListBoxItem);
      lstBox = (ListBox)ItemsControl.ItemsControlFromItemContainer(lstItem);
      itemIndex = lstBox.ItemContainerGenerator.IndexFromContainer(lstItem);

      if ((itemIndex % 2) != 0)
        ret = new Thickness(5, 0, 30, 30);
      else
        ret = new Thickness(30, 25, 0, 0);

      return ret;
    }

    public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture)
    {
      throw new NotImplementedException();
    }
  }
}

File: _wpf/wpf-business-programmer/2-wpf-business-programmer-m2-exercise-files/Samples/VB/WPFListBox-VB/WPFComponents/Converters/IndexToMarginConverter.vb
Imports System.Globalization
Imports System.Windows
Imports System.Windows.Controls
Imports System.Windows.Data

<ValueConversion(GetType(ListBoxItem), GetType(Thickness))> _
 Public Class IndexToMarginConverter
  Implements IValueConverter

  Public Function Convert(ByVal value As Object, _
                          ByVal targetType As Type, _
                          ByVal parameter As Object, _
                          ByVal culture As CultureInfo) As Object _
                          Implements System.Windows.Data.IValueConverter.Convert
    Dim ret As Thickness
    Dim itemIndex As Integer
    Dim lstItem As ListBoxItem
    Dim lstBox As ListBox

    lstItem = TryCast(value, ListBoxItem)
    lstBox = DirectCast(ItemsControl.ItemsControlFromItemContainer(lstItem), ListBox)
    itemIndex = lstBox.ItemContainerGenerator.IndexFromContainer(lstItem)

    If (itemIndex Mod 2) <> 0 Then
      ret = New Thickness(5, 0, 30, 30)
    Else
      ret = New Thickness(30, 25, 0, 0)
    End If

    Return ret
  End Function

  Public Function ConvertBack(ByVal value As Object, _
                              ByVal targetType As Type, _
                              ByVal parameter As Object, _
                              ByVal culture As CultureInfo) As Object _
                              Implements System.Windows.Data.IValueConverter.ConvertBack

    Throw New NotImplementedException()
  End Function
End Class
